shared/src/main/java/com/couchbase/lite/AbstractReplicatorConfiguration.java

```diff
@@ -0,0 +1,405 @@
+    static final String kC4AuthTypeOpenIDConnect = "OpenID Connect";
+    static final String kC4AuthTypeFacebook = "Facebook";
+    static final String kC4AuthTypeClientCert = "Client Cert";
+
+    /**
+     * Replicator type
+     * PUSH_AND_PULL: Bidirectional; both push and pull
+     * PUSH: Pushing changes to the target
+     * PULL: Pulling changes from the target
+     */
+    public enum ReplicatorType {
+        PUSH_AND_PULL,
+        PUSH,
+        PULL
+    }
+
+    //---------------------------------------------
+    // member variables
+    //---------------------------------------------
+    private Database database;
+    private ReplicatorType replicatorType;
+    private boolean continuous;
+    private Authenticator authenticator;
+    private Map<String, String> headers;
+    private byte[] pinnedServerCertificate;
+    private List<String> channels;
+    private List<String> documentIDs;
+    private ReplicationFilter pushFilter;
+    private ReplicationFilter pullFilter;
+
+    protected boolean readonly;
+    protected Endpoint target;
+
+    //---------------------------------------------
+    // Constructors
+    //---------------------------------------------
+
+    protected AbstractReplicatorConfiguration(@NonNull AbstractReplicatorConfiguration config) {
+        if (config == null) { throw new IllegalArgumentException("config cannot be null."); }
+
+        this.readonly = false;
+        this.database = config.database;
+        this.target = config.target;
+        this.replicatorType = config.replicatorType;
+        this.continuous = config.continuous;
+        this.authenticator = config.authenticator;
+        this.pinnedServerCertificate = config.pinnedServerCertificate;
+        this.headers = config.headers;
+        this.channels = config.channels;
+        this.documentIDs = config.documentIDs;
+        this.pullFilter = config.pullFilter;
+        this.pushFilter = config.pushFilter;
+    }
+
+    protected AbstractReplicatorConfiguration(@NonNull Database database, @NonNull Endpoint target) {
+        if (database == null) { throw new IllegalArgumentException("config cannot be null."); }
+        if (target == null) { throw new IllegalArgumentException("target cannot be null."); }
+
+        this.readonly = false;
+        this.replicatorType = ReplicatorType.PUSH_AND_PULL;
+        this.database = database;
+        this.target = target;
+    }
+
+    //---------------------------------------------
+    // Setters
+    //---------------------------------------------
+
+    /**
+     * Sets the authenticator to authenticate with a remote target server.
+     * Currently there are two types of the authenticators,
+     * BasicAuthenticator and SessionAuthenticator, supported.
+     *
+     * @param authenticator The authenticator.
+     * @return The self object.
+     */
+    @NonNull
+    public ReplicatorConfiguration setAuthenticator(@NonNull Authenticator authenticator) {
```

Comment:
This is interesting and I don't know how it works. ReplicatorConfiguration extends from AbstractReplicatorConfiguration but the methods in ReplicatorConfiguration returns an object of type ReplicatorConfiguration. This sounds like circular dependency.

Should this method return `AbstractReplicatorConfiguration` instead? BUT this means that ReplicatorConfiguration class will need to override each single method to return ReplicatorConfiguration.